README.md
# Cookbook ama-ssh-private-keys

[![Supermarket](https://img.shields.io/cookbook/v/ama-ssh-private-keys.svg?style=flat-square)](https://supermarket.chef.io/cookbooks/ama-ssh-private-keys)
[![CircleCI / master](https://img.shields.io/circleci/project/github/ama-team/cookbook-ssh-private-keys/master.svg?style=flat-square)](https://circleci.com/gh/ama-team/cookbook-ssh-private-keys/tree/master)
[![Scrutinizer](https://img.shields.io/scrutinizer/g/ama-team/cookbook-ssh-private-keys.svg?style=flat-square)](https://scrutinizer-ci.com/g/ama-team/cookbook-ssh-private-keys/)

This cookbook allows end user to install private keys for specified 
accounts.

Cookbook provides self-explanatory `ssh_private_key` resource:
 
```ruby
ssh_private_key 'id_rsa' do
  user 'jodie'
  content '-----BEGIN PRIVATE KEY-----...'
end
```

This will save provided key as `{jodie's home}/.ssh/id_rsa` with mode 
`0600`.

Cookbook also exposes `default` recipe that will take all items from
data bag with name specified as `[ama][ssh-private-keys][data-bag]`
attribute (`ssh-private-keys` by default) and simply pass them to 
resource.

If you want to use automatic public key generation and/or validation,
additional information is specified below.

**Eye-catching warning: if you are using this cookbook, you're
quite certainly working with keys providing some sensitive access. 
Ensure and double check that you minimize risk of exposure of your 
keys as much as possible. This cookbook, on it's behalf, tries to do
the same.**

## Requirements

- Chef 12+
- Ruby 2.3.0+
- Installed and available on $PATH ssh-keygen binary for validation
and public key generation (disabled by default)

## Tested against

- Debian 7.3+
- Ubuntu 14.04+
- Centos 6.0+
- Fedora 24+

Generally internals are very simple, so it should run anywhere.

## Full resource specification

```ruby
ssh_private_key 'hackerman:default' do
  id 'id_rsa' # name_property
  user 'hackerman' # required
  type 'ssh-rsa' # ssh-(rsa|dss|ed25519), ecdsa-sha2-nistp(256|384|521)
  content '' # required
  
  parent_directory '/workspace'
  mode '0600'
  
  # optional, required for validation / public key creation
  passphrase '2018' 
  
  # optional, required for public key creation
  comment 'hack-the-time'
  
  install_public_key true # defaults to false
  public_key_mode '0644'
  public_key_suffix '.pub'
  # if not specified, public key will be derived from private key
  # using ssh-keygen
  public_key ''
  
  perform_validation true # defaults to false
  
  action :create # :create/install, :delete/remove
end
```

If `install_public_key` is set to true, resource will create public key
file next to private key file. If public key is not supplied, it will 
be derived from private key (passphrase may be needed to do so).

If `perform_validation` is set to true, internal validator will make
following assertions before any file will be installed:

- Public key may be derived from supported private key
- If public key is provided, it matches generated public key
- If key type is provided, it matches type recovered by ssh-keygen

All not-safe-for-exposure resource properties are declared as sensitive
and won't appear in logs. Private keys are written to temporary files 
(mode 0600) for validation / public key derivation which are erased
using `ensure` blocks - only newer ssh-keygen versions accept stdin 
for reading.

# Licensing

MIT License / AMA Team


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [0.2.2] - 2018-08-23
### Changed
- Version restriction for Chef < 14 has been dropped

## [0.2.1] - 2017-08-09
### Fixed
- Fixed bug that resulted in validation error when trying to delete
key without setting it's content

## [0.2.0] - 2017-06-30
### Changed
- **[BC]** Revamped resource specification, now it has different fields
and may generate public key itself

## [0.1.1] - 2017-06-23
### Changed
- Fixed two bugs preventing default recipe from normal execution

## [0.1.0] - 2017-06-20
### Added
- Initial implementation


resources/ssh_private_key.rb
# frozen_string_literal: true

# rubocop:disable Metrics/BlockLength
# rubocop:disable Metrics/AbcSize

require_relative '../libraries/resource_helpers'
require_relative '../libraries/validator'

resource_name :ssh_private_key
default_action :create

property :id, String, name_property: true
property :user, [String, Symbol], required: true
property :content, String, required: false, sensitive: true

ssh_types = %w[rsa dsa dss].map { |type| "ssh-#{type}" }
ecdsa_types = %w[256 384 521].map { |length| "ecdsa-sha2-nistp#{length}" }
types = [*ssh_types, *ecdsa_types].reduce([]) do |carrier, type|
  carrier.push(type, type.to_sym)
end
variants = [*types, nil]
property :type, [String, Symbol, NilClass], equal_to: variants, required: false

property :mode, String, default: '0600'
property :parent_directory, [String, NilClass], required: false

property :passphrase, [String, NilClass], required: false, sensitive: true
property :comment, [String, NilClass], required: false

property :install_public_key, [TrueClass, FalseClass], default: false
property :public_key_suffix, String, default: '.pub'
property :public_key_mode, String, default: '0644'
property :public_key, [String, NilClass], required: false

property :perform_validation, [TrueClass, FalseClass], default: false

action_class do
  include AMA::Chef::SSHPrivateKeys::ResourceHelpers

  def create
    data = new_resource
    unless data.content
      raise ::Chef::Exceptions::ValidationFailed, 'content is required'
    end
    key_pair = provided_key_pair

    if data.perform_validation
      ::AMA::Chef::SSHPrivateKeys::Validator.new.validate!(key_pair)
    end

    create_key_directory

    file private_key_path do
      content data.content
      owner data.user
      mode data.mode
      sensitive true
    end

    return unless data.install_public_key

    content = (key_pair.compute_public_key || generated_public_key!).to_s
    file public_key_path do
      content content
      owner data.user
      mode data.public_key_mode
      sensitive true
      action :create
    end
  end

  def delete
    file private_key_path do
      action :delete
    end
    file public_key_path do
      action :delete
    end
  end
end

action :create do
  create
end

action :install do
  create
end

action :delete do
  delete
end

action :remove do
  delete
end
